### PR TITLE
After initial fetch, save potentially-redirected url

### PR DESCRIPTION
**Pull request by thatch**
# Pull Request Check List

This is a draft -- no tests or relnotes.  But appears to fix the issue in our environment.

Resolves: #9039

- [ ] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

### CODE REVIEW

File: src/poetry/inspection/lazy_wheel.py [hunk 1]
```diff
@@ -383,8 +383,10 @@ def _content_length_from_head(self) -> int:
         :raises HTTPRangeRequestUnsupported: if the response fails to indicate support
                                              for "bytes" ranges."""
         self._request_count += 1
-        head = self._session.head(self._url, headers=self._uncached_headers())
+        head = self._session.head(self._url, headers=self._uncached_headers(), allow_redirects=True)
         head.raise_for_status()
+        # Redirects may change this
```
Comment (dimbleby) [inline src/poetry/inspection/lazy_wheel.py:388]: does that matter?  is this necessary?  won't we just get another redirect next time and follow that too?

File: src/poetry/inspection/lazy_wheel.py [hunk 2]
```diff
@@ -569,7 +571,7 @@ def _try_initial_chunk_request(
         logger.debug("initial bytes request: %s", headers["Range"])
 
         self._request_count += 1
-        tail = self._session.get(self._url, headers=headers, stream=True)
+        tail = self._session.get(self._url, headers=headers, stream=True, allow_redirects=True)
```
[the remark this answers is not in the document]
Comment (abn) [reply src/poetry/inspection/lazy_wheel.py:574]: I agree, not a required change for this one.

File: src/poetry/inspection/lazy_wheel.py (hunk 1 above)
Comment (abn) [reply src/poetry/inspection/lazy_wheel.py:388]: I agree, not required for this change.

File: src/poetry/inspection/lazy_wheel.py [hunk 3]
```diff
@@ -383,8 +383,10 @@ def _content_length_from_head(self) -> int:
         :raises HTTPRangeRequestUnsupported: if the response fails to indicate support
                                              for "bytes" ranges."""
         self._request_count += 1
-        head = self._session.head(self._url, headers=self._uncached_headers())
+        head = self._session.head(self._url, headers=self._uncached_headers(), allow_redirects=True)
         head.raise_for_status()
+        # Redirects may change this
+        self.url = head.url
```
Comment (abn) [inline src/poetry/inspection/lazy_wheel.py:389]:
```suggestion
```

File: src/poetry/inspection/lazy_wheel.py [hunk 4]
```diff
@@ -590,6 +592,9 @@ def _try_initial_chunk_request(
                 f"did not receive partial content: got code {code}"
             )
 
+        # Redirects may change this
+        self.url = tail.url
+
```
Comment (abn) [inline src/poetry/inspection/lazy_wheel.py:597]:
```suggestion
```

**Comment by radoering** [comment]
Regarding tests: The only meaningful test I can imagine is a network test like

```
@pytest.mark.network
def test_metadata_from_wheel_url_with_redirect() -> None:
    url = (
        "https://httpbin.org/redirect-to?url=https://files.pythonhosted.org"
        "/packages/2d/99/6b0c5fe90e247b2b7b96a27cdf39ee59a02aab3c01d7243fc0c63cd7fb73"
        "/poetry_core-1.5.0-py3-none-any.whl"
    )
    metadata = metadata_from_wheel_url("poetry-core", url, requests.Session())

    assert metadata["name"] == "poetry-core"
    assert metadata["version"] == "1.5.0"
    assert metadata["author"] == "Sébastien Eustace"
    assert metadata["requires_dist"] == [
        'importlib-metadata (>=1.7.0) ; python_version < "3.8"'
    ]
```

This takes 1-2 seconds. Not sure if we want that or rather do without a test.

**Comment by abn** [comment]
@thatch please run the following locally to get the failing pre-commit job passing.

```
poetry run pre-commit install
poetry run pre-commit run --all
```

**Comment by abn** [comment]
> Regarding tests: The only meaningful test I can imagine is a network test like
> 
> ```
> @pytest.mark.network
> def test_metadata_from_wheel_url_with_redirect() -> None:
>     url = (
>         "https://httpbin.org/redirect-to?url=https://files.pythonhosted.org"
>         "/packages/2d/99/6b0c5fe90e247b2b7b96a27cdf39ee59a02aab3c01d7243fc0c63cd7fb73"
>         "/poetry_core-1.5.0-py3-none-any.whl"
>     )
>     metadata = metadata_from_wheel_url("poetry-core", url, requests.Session())
> 
>     assert metadata["name"] == "poetry-core"
>     assert metadata["version"] == "1.5.0"
>     assert metadata["author"] == "Sébastien Eustace"
>     assert metadata["requires_dist"] == [
>         'importlib-metadata (>=1.7.0) ; python_version < "3.8"'
>     ]
> ```
> 
> This takes 1-2 seconds. Not sure if we want that or rather do without a test.

We can modify the test at https://github.com/python-poetry/poetry/pull/9051/files#diff-d86ccfe236884d03286aee144a8a8dcb2016aa497261e05d81da2a7a1101ade6R292-R299 to account for this.

```py
def test_metadata_from_wheel_url_with_redirect_after_500(
    assert_metadata_from_wheel_url: AssertMetadataFromWheelUrl,
) -> None:
    assert_metadata_from_wheel_url(
        negative_offset_error=(codes.internal_server_error, b"Internal server error"),
        expected_requests=10,
        redirect=True,
    )
```

**Pull request closed without merging**